### DISCUSSION TITLE
What is the best port to choose in dfx replica to avoid port clash?

### DISCUSSION THREAD

**Discussion by q2333gh** [dfinity/sdk#3489] [Archive]
I have search the web says : <1024 is for core services. <49152  is reserved by IANA  . So best choose between 49152 through 65535 ?

`  /home/btwl/.config/dfx/networks.json `
```json
{

    "local": {
      "bind": "127.0.0.1:50010",
      "type": "ephemeral",
      "replica": {
        "subnet_type": "system"
      }
    },
    "ic": {
      "providers": ["https://ic0.app/"],
      "type": "persistent"
    },
    "ic_test": {
      "providers": ["https://ic0.app/"],
      "type": "persistent"
    },
    "web_test": {
      "providers": ["https://ic0.app/"],
      "type": "persistent"
    }

}
```

**Comment by ghost** [comment]
It turns out that using a port between 49152 and 65535 doesn't always work, because those ports can be used as ephemeral ports.  In other words, `dfx start` might work today because port 50010 happens to be available, but fail tomorrow because some other process on your system is using the port.

See [https://www.rfc-editor.org/rfc/rfc6335.html section 8.1.2](https://www.rfc-editor.org/rfc/rfc6335.html#section-8.1.2), specifically:
> On the other hand, application software MUST NOT
      assume that a specific port number in the Dynamic Ports range will
      always be available for communication at all times, and a port
      number in that range hence MUST NOT be used as a service
      identifier.

Here are two options:
1. Use an unassigned User Port (between 1024 and 49151), which can be found at https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml
2. Use a Dynamic Port, by specifying port `0` in the bind address:

```
$ cat ~/.config/dfx/networks.json 
{
  "local": {
    "bind": "127.0.0.1:0"
  }
}
$ dfx start --background 
Running dfx start for version 0.16.1
Using shared network 'local' defined in /Users/ericswanson/.config/dfx/networks.json
dfx info webserver-port
Initialized replica.
Dashboard: http://localhost:57672/_/dashboard
$ dfx info webserver-port
57668
```

**Reply by q2333gh** [reply]
Thank you so much for detailed and expanded explanation. 
Make a pretty clean and clear answer about what port to use for a regular program.
And I really like the way you answer the question and try to learn the reply method from in it !
